File: c_examples/binary_search.c
#include <stdio.h>

int binarySearch(int * list, int numItems, int valToFind)
{
    int start =0;
    int finish = numItems-1;

    while(start <= finish)
    {
        int mid = (finish + start)/2;
        if(list[mid] < valToFind)
            start = mid+1;
        else if(list[mid] > valToFind)
            finish = mid-1;
        else
            return mid;
    }

    return -1;

}

void printArray(int * list, int numItems) 
{
    printf("List :");
    for(int i=0; i<numItems;i++)
    {
        printf(" %d", list[i]);
    }
    printf("\n");
}

int main()
{
    int numList[] = { 1,2,3,4,5,6,7,8,9,10 };
    int len = sizeof(numList)/sizeof(numList[0]);
    printArray(numList, len);
    int valToFind =11;
    int index = binarySearch(numList, len, valToFind);
   
    if (index >=0)
        printf("\n %d found at pos %d\n", valToFind, index);
    else
        printf("\n %d not found\n", valToFind);


}

File: python_samples/binary_search.py
#!/usr/bin/python3

def binary_search(items, valToFind):
    sorted_items = items[:]
    sorted_items.sort()
    start = 0
    end = len(items)-1

    while start <= end:
        mid = (end + start)//2
        if sorted_items[mid] < valToFind:
            start = mid+1
        elif sorted_items[mid] > valToFind:
            end = mid-1
        else:
            return mid

    return -1

if __name__ == "__main__":
    items = [2,3,4,6,5,9,10,11]
    val = 10
    index = binary_search(items,val)
    print(items)
    if index >=0 :
        print(str(val) + " found at " + str(index))
    else:
        print(str(val) + " not found")
    
